plugins/catalog-backend-module-ldap/src/ldap/client.ts

```diff
@@ -147,39 +147,53 @@ export class LdapClient {
   async searchStreaming(
     dn: string,
     options: SearchOptions,
-    f: (entry: SearchEntry) => void,
+    f: (entry: SearchEntry) => Promise<void> | void,
   ): Promise<void> {
     try {
       return await new Promise<void>((resolve, reject) => {
         // Note that we clone the (frozen) options, since ldapjs rudely tries to
         // overwrite parts of them
-        this.client.search(dn, cloneDeep(options), (err, res) => {
-          if (err) {
-            reject(new Error(errorString(err)));
-          }
-
-          res.on('searchReference', () => {
-            this.logger.warn('Received unsupported search referral');
-          });
-
-          res.on('searchEntry', entry => {
-            f(entry);
-          });
-
-          res.on('error', e => {
-            reject(new Error(errorString(e)));
-          });
-
-          res.on('end', r => {
-            if (!r) {
-              throw new Error('Null response');
-            } else if (r.status !== 0) {
-              throw new Error(`Got status ${r.status}: ${r.errorMessage}`);
-            } else {
-              resolve();
+        this.client.search(
+          dn,
+          merge(cloneDeep(options), { paged: { pagePause: true } }),
+          (err, res) => {
+            if (err) {
+              reject(new Error(errorString(err)));
             }
-          });
-        });
+            let awaitList: Array<Promise<void> | void> = [];
+
+            res.on('searchReference', () => {
+              this.logger.warn('Received unsupported search referral');
+            });
+
+            res.on('searchEntry', entry => {
+              awaitList.push(f(entry));
+            });
+
+            res.on('page', (_, cb) => {
+              // awaits completion before fetching next page
+              Promise.allSettled(awaitList).then(() => {
```

Comment:
Man, I'm not sure about this. As a consumer of this functionality, it would be very convenient if i called the `searchStreaming` method and the promise rejected with like ```new Error(`Transform function threw an exception, ${stringifyError(e)}`)``` or whatnot, and stopped producing output at that point. So maybe additionally have a boolean flag `failed` inside here that gets set, and when it's on we no longer stream out anything?

Reply:
I just think it's maybe a bit unexpected to get (1) silent errors, and (2) randomly partial output datasets